### PR TITLE
NIFI-4774: Allow user to choose which write-ahead log implementation …

### PR DESCRIPTION
**Pull request by markap14**
…should be used by the WriteAheadFlowFileRepository

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/repository/WriteAheadFlowFileRepository.java [hunk 1]
```diff
@@ -80,6 +80,15 @@
  */
 public class WriteAheadFlowFileRepository implements FlowFileRepository, SyncListener {
     private static final String FLOWFILE_REPOSITORY_DIRECTORY_PREFIX = "nifi.flowfile.repository.directory";
+    private static final String WRITE_AHEAD_LOG_IMPL = "nifi.flowfile.repository.wal.implementation";
+
+    // TODO: Update Admin Guide
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/repository/WriteAheadFlowFileRepository.java]: I think this TODO is DONE :)

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/repository/WriteAheadFlowFileRepository.java [hunk 2]
```diff
@@ -129,17 +138,22 @@ public WriteAheadFlowFileRepository() {
         checkpointDelayMillis = 0l;
         numPartitions = 0;
         checkpointExecutor = null;
-        flowFileRepositoryPath = null;
+        walImplementation = null;
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/repository/WriteAheadFlowFileRepository.java]: Should this be DEFAULT_WAL_IMPLEMENTATION?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/repository/WriteAheadFlowFileRepository.java (hunk 1 above)
Comment (markap14) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/repository/WriteAheadFlowFileRepository.java]: Yup, this is done :) If you get a chance to merge, please delete the comment.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/repository/WriteAheadFlowFileRepository.java (hunk 2 above)
Comment (markap14) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/repository/WriteAheadFlowFileRepository.java]: It doesn't really matter what it's set to - that constructor will only be used for service loading and no method in the class will ever be called.

**Comment by mosermw** [comment]
I tested this and I was able to switch back and forth between MinimalLockingWriteAheadLog and SequentialAccessWriteAheadLog. +1 from me.
It's best to make this switch while there are 0 flowfiles in the repository.  With flowfiles in the system, going from MinimalLockingWAL to SequantialAccessWAL worked, but the opposite had some issues.

**Comment by mosermw** [comment]
@mattyb149 and @markap14 with the 1.6.0 release approaching, do you think we've allowed enough time for review and testing?  Is this good to go?  Thanks.

**Comment by mattyb149** [comment]
+1 LGTM, I tested switching back and forth and everything looks good. Thanks for the improvement! Merging to master

**Pull request closed without merging**